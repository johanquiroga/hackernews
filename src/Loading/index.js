import React from 'react';
import FontAwesomeIcon from '@fortawesome/react-fontawesome';

const Loading = () => <FontAwesomeIcon icon="spinner" spin />

export default Loading;
