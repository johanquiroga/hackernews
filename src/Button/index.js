import React from 'react';
import PropTypes from 'prop-types';
import { withLoading } from '../HOC';

const Button = ({ onClick, className, children }) =>
  <button
    onClick={onClick}
    className={className}
    type="button"
  >
    {children}
  </button>

Button.defaultProps = {
  className: '',
};

Button.propTypes = {
  onClick: PropTypes.func.isRequired,
  className: PropTypes.string,
  children: PropTypes.node.isRequired,
};

export default Button;

export const ButtonWithLoading = withLoading(Button);
