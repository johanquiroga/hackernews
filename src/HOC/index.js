import React from 'react';
import Loading from '../Loading';
import Alert from '../Alert';
import Button from '../Button';

export const withLoading = (Component) => (props) =>
  <div>
    <Component { ...props } />

    <div className="interactions">
      {props.isLoading && <Loading />}
    </div>
  </div>

export const withPaginated = (Component) => (props) =>
  <div>
    <Component {...props} />

    <div className="interactions">
      {
        (props.page !== null && !props.isLoading) &&
        <Button
          onClick={props.onPaginatedSearch}
        >
          More
        </Button>
      }
    </div>
  </div>

export const withError = (Component) => ({error, ...rest}) =>
  error
  ? <Alert msg="Something went wrong!" />
  : <Component { ...rest } />
