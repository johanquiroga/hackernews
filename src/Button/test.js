import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import Enzyme, { shallow, render, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { compose } from 'recompose';

import Button from './index.js';
import Loading from '../Loading';
import { withLoading } from '../HOC';

import fontawesome from '@fortawesome/fontawesome';
import FontAwesomeIcon from '@fortawesome/react-fontawesome';
import { faSpinner, faSort, faSortUp, faSortDown } from '@fortawesome/fontawesome-free-solid';

Enzyme.configure({ adapter: new Adapter() });

describe('Button', () => {
  const props = {
    className: 'button',
    onClick() {console.log('Button clicked')}
  };

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Button { ...props }>Give me More</Button>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('sets the given class', () => {
    const element = render(<Button { ...props }>Give me More</Button>);
    expect(element.is('.button')).toBe(true);
  });

  it('sets the correct child', () => {
    const element = render(<Button { ...props }>Give me More</Button>);
    expect(element.text()).toEqual('Give me More');
  });

  test('has a valid snapshot', () => {
    const component = renderer.create(<Button { ...props }>Give me More</Button>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('ButtonWithLoading', () => {
  const props = {
    onClick() {console.log('Button clicked')}
  };

  const ButtonWithLoading = compose(
    withLoading
  )(Button);

  it('renders button without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<ButtonWithLoading isLoading={false} {...props}>More</ButtonWithLoading>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders Loading without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<ButtonWithLoading isLoading={true} {...props}>More</ButtonWithLoading>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('mounts a Button Component', () => {
    const component = mount(<ButtonWithLoading isLoading={false} {...props}>More</ButtonWithLoading>);
    expect(component.find(Button).length).toBe(1);
  });

  it('mounts a Loading Component', () => {
    const component = mount(<ButtonWithLoading isLoading={true} {...props}>More</ButtonWithLoading>);
    expect(component.find(Loading).length).toBe(1);
  });

  test('has a valid button snapshot', () => {
    const component = renderer.create(<ButtonWithLoading isLoading={false} {...props}>More</ButtonWithLoading>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('has a valid loading snapshot', () => {
    const component = renderer.create(<ButtonWithLoading isLoading={true} {...props}>More</ButtonWithLoading>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
