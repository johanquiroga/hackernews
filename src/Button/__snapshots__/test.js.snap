// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button has a valid snapshot 1`] = `
<button
  className="button"
  onClick={[Function]}
  type="button"
>
  Give me More
</button>
`;

exports[`ButtonWithLoading has a valid button snapshot 1`] = `
<div>
  <button
    className=""
    onClick={[Function]}
    type="button"
  >
    More
  </button>
  <div
    className="interactions"
  />
</div>
`;

exports[`ButtonWithLoading has a valid loading snapshot 1`] = `
<div>
  <button
    className=""
    onClick={[Function]}
    type="button"
  >
    More
  </button>
  <div
    className="interactions"
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-spinner fa-w-16 fa-spin "
      data-icon="spinner"
      data-prefix="fas"
      role="img"
      style={Object {}}
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M304 48c0 26.51-21.49 48-48 48s-48-21.49-48-48 21.49-48 48-48 48 21.49 48 48zm-48 368c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm208-208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zM96 256c0-26.51-21.49-48-48-48S0 229.49 0 256s21.49 48 48 48 48-21.49 48-48zm12.922 99.078c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.491-48-48-48zm294.156 0c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.49-48-48-48zM108.922 60.922c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.491-48-48-48z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
  </div>
</div>
`;
